reviews/2020-openebs-annual.md

```diff
@@ -0,0 +1,166 @@
+  * https://github.com/kubernetes-csi/
+  * https://github.com/kubernetes-sigs
+  * https://github.com/spdk
+  * https://github.com/grpc
+  * https://github.com/vmware-tanzu/velero
+  * https://github.com/longhorn/
+  * https://github.com/gostor/gotgt
+
+## Alignment with Cloud Native
+
+OpenEBS falls within the scope of they [CNCF Storage SIG](https://github.com/cncf/sig-storage/blob/master/storage-charter.md#areas-considered-in-scope), offering a significantly different approach to implementing a block storage system by making use of cloud-native principles and Kubernetes itself as a key component to delivering storage for application workloads.
+
+OpenEBS and its community are aligned with the CNCF mission.
+
+- Extensible and open architecture based upon Kubernetes. Built with Kubernetes, for Kubernetes. 
+- Horizontally scalable with Kubernetes nodes allowing deployments from small developer-centric to large scale enterprise environments. 
+- Works well and users have deployed OpenEBS alongside other CNCF projects like Kubernetes, Prometheus, etcd, and others
+- Additionally, the OpenEBS community attempts to be extremely inclusive and is one of the more active CNCF related projects in India, actively working with technical universities to introduce interested students to Open Source practices and projects.
+
+### Comparison with Longhorn
+Longhorn is another CNCF project. One of the OpenEBS Storage engines, Jiva, was forked out of Longhorn-engine and the OpenEBS community has kept the engine fixes as compatible as possible. 
+
+OpenEBS Jiva engine differs from Longhorn:
+
+- Deployment and control plane operations. OpenEBS Jiva engine continues to provide a complete isolated set of containers for each volume and is designed to self-heal from cluster failures or other chaos. Longhorn currently depends on the control plane pods for managing and recovering from failures.  
+- Jiva remains focused on providing volumes for workloads that are short-lived or have low capacity requirements. This allows Jiva to not require inbuilt support for scheduled snapshots and restore. However, Jiva does support the API for administrators to take remote backups for handling complete cluster migration. 
+
+In addition to Jiva, other storage engines included as a part of OpenEBS include:
+
+- LocalPV - including hostpath, device, and ZFS
+- cStor - including an efficient COW engine for snapshots and clones
+- MayaStor - a soon to be beta engine refactored in Rust and intended to be a high performance engine, with native use of NVMe
+  
+### Comparison with Rook
+
+Rook is another CNCF project that focuses on:
+
+- Simplifying the management of Storage engines that are developed outside of the CNCF like Ceph and EdgeFS. 
+- Provide a framework to develop higher-level abstractions and operators for running stateful workloads like CockroachDB, Cassandra. 
+
+OpenEBS is focused on Storage engines and the standard Kubernetes abstractions to use them.  
```

Comment:
Rook is very commonly used with Ceph which is also open source. Can you provide a comparison to Rook+Ceph vs OpenEBS?

Reply:
Hi @saad-ali - Thanks for bringing this up. This is one of the common questions, though I wasn't sure if we should really include this comparison into the review documents as CEPH isn't a CNCF project. 

IMO, both are good and serve different use cases - though if may add OpenEBS is more focused on providing storage for micro-services/container-based workloads using the loosely coupled architecture as opposed to everything shared architecture. 

Expanding on this a little more: 
* In terms of architecture, I see CEPH as a good example of distributed shared scale-out storage whereas OpenEBS embraces the microservices pattern namely loosely coupled independent systems for each workload. The scale-out aspects in OpenEBS are tied into the auto-scaling patterns of Kubernetes itself.  
* In terms of usage, CEPH can provide volumes of very large capacity because of its distributed and shared scale-out architecture whereas OpenEBS is best suited for microservices workloads that need a large number of volumes with small capacities. Prior to containers, it was pretty common to provision 200GB and above volumes, but now we see a lot of workloads with 5 or 10GB. 
* In terms of scale, a single CEPH cluster is good for running workloads up to a max of certain capacity and volumes as the metadata systems are shared for all volumes, whereas a single openebs cluster can handle more volumes by auto-scaling its volume pods. 
* In terms of IO processing, CEPH goes through some shared layers for all the workloads whereas with OpenEBS - there are multiple storage engines - that are selected by storage classes within Kubernetes per workload.  For example, the cStor storage engine is typically selected for workloads that are not highly resilient and may benefit from a copy on write back-ups and snapshots whereas the LocalPV options are used for highly resilient workloads such as Cassandra and Kafka that mainly need assistance in managing the underlying storage media and in tracking inventory of available resources via CRDs.
* In terms of Access Protocols, OpenEBS is primarily for block storage.  For read-write many workloads, NFS can be deployed and this is an area of active development.  By comparison, CEPH intended to serve all workloads - naturally offers block, file, and object interfaces.
* A few other points of differentiation include: 
  - CEPH depends upon various kernel modules to deliver its functionality, OpenEBS as container attached storage attempts to be as portable as containers and will work without modification or kernel modules on any compatible Kubernetes environment given that the environment supports iSCSI - which is a widely deployed option.
  - OpenEBS includes a storage engine, Mayastor, which while early in its usage appears to be more promising than early projects within CEPH to support NVMe access to underlying media.  Here again, OpenEBS benefits from the userspace architecture which fits well with Intel's underlying SPDK project for lockless access to extremely high-performance storage media.